src/uwtools/j2template.py

```diff
@@ -0,0 +1,123 @@
+        List of variables needed given a Jinja2 template
+
+    Methods
+    -------
+
+    _load_file()
+        Wrapper around loading a Jinja2 template from a file.
+
+    _load_string()
+        Wrapper around loading a Jinja2 template from a string
+
+    dump_file(output_path)
+        Wrapper around writing the template to an output path
+
+    validate_config()
+        Checks to ensure that the provided configure_obj is sufficient
+        to meet the needs of the undeclared_variables in the template
+
+    '''
+
+    def __init__(self, configure_obj, template_path=None, template_str=None):
+
+        '''
+        Parameters
+        ----------
+        configure_obj : dict (See above)
+        template_path : Path
+            Path to a Jinja2 template file
+        template_str : str
+            A Jinja2 template string
+
+        '''
+        self.configure_obj = configure_obj
+        self.template_path = template_path
+        self.template_str = template_str
+        if template_path is not None:
+            self.template = self._load_file(template_path)
+        elif template_str is not None:
+            self.template = self._load_string(template_str)
+        else:
+            # Error here. Must provide a templateA
+            pass
+
+    def dump_file(self,output_path):
+
+        '''
+        Write rendered template to the output_path provided
+
+        Parameters
+        ----------
+        output_path : Path
+
+        '''
+        _render = self.render_template()
+        with open(output_path,'w+',encoding='utf-8') as _file:
+            _file.write(_render)
+
+    def _load_file(self, template_path):
+        '''
+        Load the Jinja2 template from the file provided.
+
+        Returns
+        -------
+        Jinja2 Template object
+        '''
+
+        self._j2env=Environment(loader=FileSystemLoader(searchpath='/'),
+                               trim_blocks=True,lstrip_blocks=True)
+        return self._j2env.get_template(template_path)
+
+    def _load_string(self,template_str):
+        '''
+        Load the Jinja2 template from the string provided.
+
+        Returns
+        -------
+        Jinja2 Template object
+        '''
+        return Environment(loader=BaseLoader()).from_string(template_str)
+
+    def render_template(self):
+        ''' Render the Jinja2 template so that it's available in memory
+
+        Returns
+        -------
+        A string containing a rendered Jinja2 template.
+        '''
+        return self.template.render(self.configure_obj)
+    
+    @property
+    def undeclared_variables(self):
+
+        ''' Generates a list of variables needed for self.template '''
+        if self.template_str is not None:
+            j2_parsed = self._j2env.parse(self.template_str)
+        else:
+            with open(self.template_path,encoding='utf-8') as __file:
+                j2_parsed = self._j2env.parse(__file.read())
+        return meta.find_undeclared_variables(j2_parsed)
```

Comment:
As we'd discussed earlier, the test for this method fails like this:
```
self = <uwtools.j2template.J2Template object at 0x7f03b872b5d0>

    @property
    def undeclared_variables(self):
    
        ''' Generates a list of variables needed for self.template '''
        if self.template_str is not None:
>           j2_parsed = self._j2env.parse(self.template_str)
E           AttributeError: 'J2Template' object has no attribute '_j2env'

.tox/py/lib/python3.7/site-packages/uwtools/j2template.py:119: AttributeError
```

That's related to the object attribute `_j2env` not being set appropriately in the case of a string template. That means this is a really nice test since it caught a limitation of the code. I'm thinking that you could modify the `_load_string` method to set the attribute and then return the template object to mirror what is done for `_load_file`. Something like this untested code might do the trick:

```
def _load_string(self):

    self._j2env = Environment(loader=BaseLoader())
    return self._j2env.from_string(template_str)
  
```

Given that we are assuming the load functions set this attribute, it might be worth a similar test for undeclared variables in the existing file loading test you already have. That should help future developers make sure that they don't refactor `_load_file` and mess up `undeclared_variables`.